agent/consul/leader_connect_test.go

```diff
@@ -396,23 +396,35 @@ func TestLeader_Vault_PrimaryCA_IntermediateRenew(t *testing.T) {
 	provider, _ := getCAProviderWithLock(s1)
 	intermediatePEM, err := provider.ActiveIntermediate()
 	require.NoError(err)
-	_, err = connect.ParseCert(intermediatePEM)
+	intermediateCert, err := connect.ParseCert(intermediatePEM)
 	require.NoError(err)
 
+	// Check that the state store has the correct intermediate
+	store := s1.caManager.delegate.State()
+	_, activeRoot, err := store.CARootActive(nil)
+	require.NoError(err)
+	require.Equal(intermediatePEM, activeRoot.LeafSigningCert())
+	require.Equal(connect.HexString(intermediateCert.SubjectKeyId), activeRoot.SigningKeyID)
+	t.Log(activeRoot.SigningKeyID)
+
 	// Wait for dc1's intermediate to be refreshed.
 	// It is possible that test fails when the blocking query doesn't return.
 	retry.Run(t, func(r *retry.R) {
 		provider, _ = getCAProviderWithLock(s1)
 		newIntermediatePEM, err := provider.ActiveIntermediate()
 		r.Check(err)
-		_, err = connect.ParseCert(intermediatePEM)
-		r.Check(err)
 		if newIntermediatePEM == intermediatePEM {
 			r.Fatal("not a renewed intermediate")
 		}
+		intermediateCert, err = connect.ParseCert(newIntermediatePEM)
+		r.Check(err)
```

Comment:
out of curiosity, why do we rely on the `*retry.R` checks here and not a `require.NoError()` call? 📚

Reply:
Good question! 

For `retry.Run` to actually retry things the failure has to be reported to `r` and not `t`, so we have to be careful to use the right thing, but that may not be your question.

If you mean why are we using `r.Check(rrr)` instead of `require.NoError(t, err)`, I would say there's no real reason. I kept what existed mostly, but they should be equivalent.

I'll change this to use `require.NoError()` so that we can probably deprecate and remove `r.Check()`

Reply:
Oh, but because we've aliased `require := require.New(t)` this no longer works. I've always found `require.New(t)` to be more of a problem than a solution. It makes it way too easy to use the wrong `testing.T` when subtests are involved.

I don't want to do major cleanup to this test in this PR, so I'll keep `r.Check` for now, and we can try to clean some of that up later.

Reply:
Thanks for the context here!

Reply:
btw, I played around w `require` calls (with a shadowed `require := require.New(t)`) inside a `retry.Run()` call and the require checks worked as expected 🤷🏼

Reply:
Hmm, it shouldn't! If `t.Fatalf()` is called (instead of `r.Fatalf()`) it would stop the test run instead of retrying again. I'm not sure why it appeared to work, maybe it passed the first time instead of failing, so the retry never happened?